### PR TITLE
feat: Add UDT for all `avm/res/sql/server` submodules

### PR DESCRIPTION
**Pull request by peterbud**
## Description

This PR adds the remaining UDTs for al SQL submodules

## Pipeline Reference

| Pipeline |
| -------- |
|     [avm.res.sql.server](https://github.com/peterbud/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml)    |

Pls note that WAF fails due to missing maintenance window in my subscription.

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [x] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/sql/server/encryption-protector/main.bicep [hunk 1]
```diff
@@ -8,10 +8,10 @@ param sqlServerName string
 @description('Required. The name of the server key.')
 param serverKeyName string
 
-@description('Optional. Key auto rotation opt-in.')
-param autoRotationEnabled bool = false
+@description('Optional. Key auto rotation opt-in flag. Either true or false.')
+param autoRotationEnabled bool = true
 
-@description('Optional. The encryption protector type.')
+@description('Optional. The encryption protector type like \'ServiceManaged\', \'AzureKeyVault\'.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/encryption-protector/main.bicep:14]:
```suggestion
@description('Optional. The encryption protector type.')
```
The allowed values are anyways provided to the user when calling the parameter

File: avm/res/sql/server/encryption-protector/main.bicep [hunk 2]
```diff
@@ -8,10 +8,10 @@ param sqlServerName string
 @description('Required. The name of the server key.')
 param serverKeyName string
 
-@description('Optional. Key auto rotation opt-in.')
-param autoRotationEnabled bool = false
+@description('Optional. Key auto rotation opt-in flag. Either true or false.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/encryption-protector/main.bicep:11]:
```suggestion
@description('Optional. Key auto rotation opt-in flag.')
```
I guess that should be a given 😉

File: avm/res/sql/server/key/main.bicep [hunk 3]
```diff
@@ -8,14 +8,14 @@ param name string?
 @description('Conditional. The name of the parent SQL server. Required if the template is used in a standalone deployment.')
 param serverName string
 
-@description('Optional. The encryption protector type like "ServiceManaged", "AzureKeyVault".')
+@description('Optional. The server key type like \'ServiceManaged\', \'AzureKeyVault\'.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/key/main.bicep:11]:
```suggestion
@description('Optional. The server key type.')
```
See previous comment.

File: avm/res/sql/server/main.bicep [hunk 4]
```diff
@@ -799,3 +798,96 @@ type elasticPoolPropertyType = {
...
+
+@export()
+type encryptionProtectorType = {
+  @description('Required. The name of the server key.')
+  serverKeyName: string
+
+  @description('Optional. The encryption protector type like \'ServiceManaged\', \'AzureKeyVault\'.')
+  serverKeyType: 'ServiceManaged' | 'AzureKeyVault'?
+
+  @description('Optional. Key auto rotation opt-in flag. Either true or false.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:810]:
```suggestion
  @description('Optional. Key auto rotation opt-in flag.')
```
A given for type boolean.

File: avm/res/sql/server/main.bicep [hunk 5]
```diff
@@ -799,3 +798,96 @@ type elasticPoolPropertyType = {
...
+  @description('Optional. The end IP address of the firewall rule. Must be IPv4 format. Must be greater than or equal to startIpAddress. Use value \'0.0.0.0\' for all Azure-internal IP addresses.')
+  endIpAddress: string?
+}
+
+@export()
+type keyType = {
+  @description('Optional. The name of the key. Must follow the [<keyVaultName>_<keyName>_<keyVersion>] pattern.')
+  name: string?
+
+  @description('Optional. The server key type like \'ServiceManaged\', \'AzureKeyVault\'.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:849]:
```suggestion
  @description('Optional. The server key type.')
```
See previous comment.

File: avm/res/sql/server/main.bicep [hunk 6]
```diff
@@ -557,6 +555,7 @@ output privateEndpoints array = [
 import { diagnosticSettingFullType } from 'br/public:avm/utl/types/avm-common-types:0.2.1'
```
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:555]: As a recommendation as per the other type imports from the avm-common-types, please move this import on top of the parameter where it is used. Makes it a bit easier to understand where the type is coming from.
The reamining types are fine were they are. Would just be nice to have the consistency for the common types.

File: avm/res/sql/server/security-alert-policy/main.bicep [hunk 7]
```diff
@@ -6,13 +6,13 @@ metadata owner = 'Azure/module-maintainers'
 param name string
 
 @description('Optional. Specifies an array of alerts that are disabled. Allowed values are: Sql_Injection, Sql_Injection_Vulnerability, Access_Anomaly, Data_Exfiltration, Unsafe_Action, Brute_Force.')
-param disabledAlerts array = []
+param disabledAlerts string[] = []
```
Comment (AlexanderSehr) [inline avm/res/sql/server/security-alert-policy/main.bicep:9]: As per the description, this should be changed to
```bicep
@description('Optional. Alerts to disable.')
@allowed([
  'Sql_Injection'
  'Sql_Injection_Vulnerability'
  'Access_Anomaly'
  'Data_Exfiltration'
  'Unsafe_Action'
  'Brute_Force'
])
param disabledAlerts string[] = []
```

File: avm/res/sql/server/main.bicep [hunk 8]
```diff
@@ -799,3 +798,96 @@ type elasticPoolPropertyType = {
...
+  @description('Optional. Allow creating a firewall rule before the virtual network has vnet service endpoint enabled.')
+  ignoreMissingVnetServiceEndpoint: bool?
+}
+
+@export()
+type securityAlerPolicyType = {
+  @description('Required. The name of the Security Alert Policy.')
+  name: string
+
+  @description('Optional. Specifies an array of alerts that are disabled. Allowed values are: Sql_Injection, Sql_Injection_Vulnerability, Access_Anomaly, Data_Exfiltration, Unsafe_Action, Brute_Force.')
```
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:873]:
```suggestion
  @description('Optional. The alerts to disable.')
  disabledAlerts: ('Sql_Injection' | 'Sql_Injection_Vulnerability' | 'Access_Anomaly' | 'Data_Exfiltration' | 'Unsafe_Action' | 'Brute_Force')[]?
```

File: avm/res/sql/server/encryption-protector/main.bicep (hunk 2 above)
Comment (peterbud) [reply avm/res/sql/server/encryption-protector/main.bicep:11]: Agree. My principle was always to take the original description of the RP (in this case [here](https://learn.microsoft.com/en-us/azure/templates/microsoft.sql/servers/encryptionprotector?pivots=deployment-language-bicep#encryptionprotectorproperties)), but we can simplify those messages in similar cases.

File: avm/res/sql/server/main.bicep (hunk 6 above)
Comment (peterbud) [reply avm/res/sql/server/main.bicep:555]: It is not used at a parameter, but as a type _within_ the `databasePropertyType`. You cannot move a type import _inside_ a type declaration AFAIK. What would be your suggestion for that?

Comment (AlexanderSehr) [reply avm/res/sql/server/main.bicep:555]: In that case nevermind :) The error was on my part. Thanks for elaborating @peterbud

**Comment by peterbud** [comment]
Anything else needed to merge this PR? LMK

**Comment by AlexanderSehr** [comment]
> Anything else needed to merge this PR? LMK

Just to get a green workflow run ;) I'd recommend to trigger the 're-run failed jobs' on the latest workflow run and only if anything still fails, look deeper into why (i.e., check if there's a bug)

**Comment by peterbud** [comment]
> Just to get a green workflow run ;) I'd recommend to trigger the 're-run failed jobs' on the latest workflow run and only if anything still fails, look deeper into why (i.e., check if there's a bug)

Fair enough. I have updated the above badge with the latest run.

Pls note that WAF fails due to missing maintenance window in my subscription.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**